zcash_keys/src/address.rs

```diff
@@ -168,43 +186,84 @@ impl UnifiedAddress {
+    }
+
+    /// Returns the expiration height for this address.
+    pub fn expiry_height(&self) -> Option<BlockHeight> {
+        self.expiry_height
+    }
+
+    /// Returns the expiration time for this address as a Unix Epoch Time.
+    pub fn expiry_time(&self) -> Option<u64> {
+        self.expiry_time
+    }
+
+    /// Returns any unknown metadata items parsed from the encoded form of the address.
+    ///
+    /// Unknown metadata items are guaranteed by construction and parsing to not have keys in the
+    /// MUST-understand metadata typecode range.
+    pub fn unknown_metadata(&self) -> &[(u32, Vec<u8>)] {
+        self.unknown_metadata.as_ref()
     }
 
     fn to_address(&self, net: NetworkType) -> ZcashAddress {
-        let items = self
-            .unknown
-            .iter()
-            .map(|(typecode, data)| unified::Receiver::Unknown {
-                typecode: *typecode,
-                data: data.clone(),
-            });
+        let data_items =
+            self.unknown_data
+                .iter()
+                .map(|(typecode, data)| unified::Receiver::Unknown {
+                    typecode: *typecode,
+                    data: data.clone(),
+                });
 
         #[cfg(feature = "orchard")]
-        let items = items.chain(
+        let data_items = data_items.chain(
             self.orchard
                 .as_ref()
                 .map(|addr| addr.to_raw_address_bytes())
                 .map(unified::Receiver::Orchard),
         );
 
         #[cfg(feature = "sapling")]
-        let items = items.chain(
+        let data_items = data_items.chain(
             self.sapling
                 .as_ref()
                 .map(|pa| pa.to_bytes())
                 .map(unified::Receiver::Sapling),
         );
 
-        let items = items.chain(self.transparent.as_ref().map(|taddr| match taddr {
+        let data_items = data_items.chain(self.transparent.as_ref().map(|taddr| match taddr {
             TransparentAddress::PublicKeyHash(data) => unified::Receiver::P2pkh(*data),
             TransparentAddress::ScriptHash(data) => unified::Receiver::P2sh(*data),
         }));
 
-        let ua = unified::Address::try_from_items(items.collect())
-            .expect("UnifiedAddress should only be constructed safely");
+        let meta_items = self
+            .unknown_metadata
+            .iter()
+            .map(|(typecode, data)| unified::MetadataItem::Unknown {
+                typecode: *typecode,
+                data: data.clone(),
+            })
+            .chain(
+                self.expiry_height
+                    .map(|h| unified::MetadataItem::ExpiryHeight(u32::from(h))),
+            )
+            .chain(self.expiry_time.map(unified::MetadataItem::ExpiryTime));
+
+        let ua = unified::Address::try_from_items(
+            if self.expiry_height().is_some() || self.expiry_time().is_some() {
+                Revision::R1
+            } else {
+                Revision::R0
```

Comment:
This causes the construction to fail for a transparent-only UA without expiry height or time. That's fine, but we should enforce this policy consistently.

Reply:
Fixed in 078b4ffa4577c8cb0e1a325b6a1b2b512c20a18a